File: Python3/1962_remove_stones_to_minimize_the_total.py
# https://leetcode.cn/problems/remove-stones-to-minimize-the-total/

class Solution:
    def minStoneSum(self, piles, k):
        """
        Inputs:
            piles: list[int]
            k: int
        Outputs:
            res: int
        """
        pq = []
        for pile in piles:
            heapq.heappush(pq, -pile)
        
        for _ in range(k):
            pile = -heapq.heappop(pq)
            heapq.heappush(pq, -((pile+1)//2))
        return -sum(pq)

File: Cpp/1962_remove_stones_to_minimize_the_total.cpp
// https://leetcode.cn/problems/remove-stones-to-minimize-the-total/

class Solution {
public:
    int minStoneSum(vector<int>& piles, int k) {
        priority_queue<int> pq;
        for (int pile: piles) {
            pq.push(pile);
        }
        for (int i = 0; i < k; ++i) {
            int pile = pq.top();
            pq.pop();
            pq.push((pile+1) / 2);
        }
        int res = 0;
        while (!pq.empty()) {
            res += pq.top();
            pq.pop();
        }
        return res;
    }
};
